analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs

```diff
@@ -0,0 +1,124 @@
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[ExportCodeFixProvider(LanguageNames.CSharp)]
+public sealed class CommentedOutCodeCodeFix : SonarCodeFix
+{
+    public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(CommentedOutCode.DiagnosticId);
+
+    public override FixAllProvider GetFixAllProvider() => null;
+
+    protected override Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext context)
+    {
+        var diagnostic = context.Diagnostics.First();
+        var comment = root.FindTrivia(diagnostic.Location.SourceSpan.Start, findInsideTrivia: true);
+
+        if (comment.IsKind(SyntaxKind.SingleLineCommentTrivia) || IsCode(comment))
+        {
+            context.RegisterCodeFix(CodeAction.Create(
+                CommentedOutCode.MessageFormat,
+                c => new Context(comment).ChangeDocument(context.Document, root)),
+                diagnostic);
+        }
+        return Task.CompletedTask;
+    }
+
+    private static bool IsCode(SyntaxTrivia trivia) =>
+        trivia.GetLineNumbers().Count() == 1
+        || Array.TrueForAll(Lines(trivia), IsCode);
+
+    private static bool IsCode(string line)
+    {
+        var trimmed = line.Trim();
+        return trimmed == "/*"
+            || trimmed == "*/"
+            || string.IsNullOrEmpty(trimmed)
+            || CommentedOutCode.IsCode(line);
+    }
+
+    private static string[] Lines(SyntaxTrivia trivia) =>
+        trivia.ToFullString().Split(MetricsBase.LineTerminators, StringSplitOptions.None);
+
+    internal sealed class Context
+    {
+        public Context(SyntaxTrivia comment)
+        {
+            Comment = comment;
+            Leading = Token.LeadingTrivia.ToList();
+            Trailing = Token.TrailingTrivia.ToList();
+
+            IsTrailing = Trailing.Remove(comment);
+
+            if (!IsTrailing)
+            {
+                Leading.Remove(comment);
+            }
+        }
+
+        public bool IsTrailing { get; }
+        public SyntaxTrivia Comment { get; }
+        public List<SyntaxTrivia> Leading { get; }
+        public List<SyntaxTrivia> Trailing { get; }
+        public SyntaxToken Token => Comment.Token;
+
+        public Task<Document> ChangeDocument(Document document, SyntaxNode root) =>
+            Task.FromResult(document.WithSyntaxRoot(root.ReplaceToken(Token, NewToken)));
+
+        public SyntaxToken NewToken =>
+            IsTrailing ? NewTrailing() : NewLeading();
+
+        private SyntaxToken NewTrailing()
+        {
+            var trailing = ShareLine(Token, Comment)
+                ? Trailing.Where(t
+                    => !ShareLine(t, Comment)
+                    || (t.IsKind(SyntaxKind.EndOfLineTrivia) && ShareLine(t, Comment))).ToList()
+                : Trailing;
+
+            // this can happen for multi line comments within an expression.
+            if (trailing.Count == 0 && ShareLine(Token.GetNextToken(), Comment))
```

Comment:
Optional: What do you think of adding a concrete example here?

```csharp
            // this can happen for multi line comments within an expression.
            // int /* return 42; */ x;
            //    ^                ^    one of the spaces is needed

```

I think it's a great catch, but I could not reason about it before running the UTs and stepping into it.
(It's kind of ironic that to add this comment, you would need to add an issue of the same rule :) )

Reply:
That case is covered with: `int /* return 17; */ WithinLine() { return 42; }`.

Reply:
I meant to add it actually in the code, yes. 
On the one hand, I think it would be nice for someone reading this code for the first time.
On the other hand, it's a code smell :)

I will leave it up to you on this one.

Reply:
I already added it.